### PR TITLE
Support for SBC in front of Restcomm Pull Request Review

### PR DESCRIPTION
**Pull request by leftyb**
#2741

### CODE REVIEW

**Comment by abdulazizali77** [comment]
@leftyb 
- Can you have a look at the failing tests, they are failing specifically in the areas you changed `UserAgentManager` 
`MonitoringServiceTool.getInstance().getRegisteredUsers`

- Please reset the branch to before the `MSB-352` cherrypick

**Comment by deruelle** [comment]
The test failures in the CI job are passing locally

File: restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/Call.java [hunk 1]
```diff
@@ -2185,6 +2196,8 @@ private void onSipServletResponse(SipServletResponse message, ActorRef self, Act
                         uri.setLrParam(true);
                         challengeRequest.pushRoute(uri);
                     }
+                    //FIXME:should check for the Ims pushed Route?
```
Comment (jaimecasero) [inline restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/Call.java]: is this FIXME applicable anymore?

File: restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/ua/UserAgentManager.java [hunk 2]
```diff
@@ -662,17 +716,30 @@ private String createAddress(SipServletRequest request) throws ServletParseExcep
         } else {
             buffer.append("sip:");
         }
-        buffer.append(normalize(user)).append("@").append(uri.getHost()).append(":").append(uri.getPort());
+        int contactPort = contactUri.getPort();
+        if(contactPort < 0) {
+            if(transport == null || transport.equalsIgnoreCase("udp") || transport.equalsIgnoreCase("tcp")) {
+                contactPort = 5060;
```
Comment (jaimecasero) [inline restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/ua/UserAgentManager.java]: where are these ports coming from, shouldnt be coming from env configuration?

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/DialActionOrganizationSBCTest.java [hunk 3]
```diff
@@ -0,0 +1,1079 @@
...
+    @ArquillianResource
+    private Deployer deployer;
+    @ArquillianResource
+    URL deploymentUrl;
+
+	@GetDeployableContainer
+	private ContainerManagerTool containerManager;
+
+    @Rule
+    public WireMockRule wireMockRule = new WireMockRule(8090); // No-args constructor defaults to port 8080
```
Comment (jaimecasero) [inline restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/DialActionOrganizationSBCTest.java]: this is not following paralle testing approach, why dont we use paralel testing here?

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/DialActionOrganizationSBCTest.java [hunk 4]
```diff
@@ -0,0 +1,1079 @@
...
+     *
+     * test case: maria calls alice.
+     *
+     * result: call do not go through
+     *
+     * @throws ParseException
+     * @throws InterruptedException
+     */
+    @Test
+    public void testClientsCallEachOtherDifferentOrganization() throws ParseException, InterruptedException {
```
Comment (jaimecasero) [inline restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/DialActionOrganizationSBCTest.java]: this should be in "exception" category IMO

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/DialActionOrganizationSBCTest.java [hunk 5]
```diff
@@ -0,0 +1,1079 @@
...
+     *
+     * test case 1: bob@org2 created INVITE - sip:+12223334467@org3.restcomm.com -> call should NOT go to alice@org3 (bcz 12223334467@org3.restcomm.com is pure sip) - instead call should FAIL
+     * test case 2: bob@org2 created INVITE - sip:+12223334467@default.restcomm.com -> call should FAIL (bcz defaulOrg does not have that number)
+     * test case 3: bob@org3 created INVITE - sip:+12223334467@org2.restcomm.com(a conference in org2) -> able to join conference (bcz 12223334467@org2.restcomm.com is provider number)
+     * @throws ParseException
+     * @throws InterruptedException
+     * @throws UnknownHostException
+     */
+    @Test
+    public void testDialNumberExistingInMultipleOrganizationCase2() throws ParseException, InterruptedException, UnknownHostException {
```
Comment (jaimecasero) [inline restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/DialActionOrganizationSBCTest.java]: this maybe an "alternate" category test case

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/ua/UserAgentManagerWithSBCTest.java [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+        logger.info("RequestEvent :"+requestEvent.getRequest().toString());
+        Response response = sipStack.getMessageFactory().createResponse(SIPResponse.OK, requestEvent.getRequest());
+        phone.sendReply(requestEvent, response);
+        Thread.sleep(1000);
+        // Clean up (Unregister).
+        assertTrue(phone.unregister(aliceContact, 0));
+    }
+    
+    @Test
+    public void registerUserAgentWith408ErrorResponse() throws ParseException, InterruptedException, InvalidArgumentException {
```
Comment (jaimecasero) [inline restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/ua/UserAgentManagerWithSBCTest.java]: "exception" category

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/ua/UserAgentManagerWithSBCTest.java [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+        assertEquals(1, MonitoringServiceTool.getInstance().getRegisteredUsers(deploymentUrl.toString(),adminAccountSid, adminAuthToken));
+
+        Thread.sleep(50000);
+        assertEquals(0, MonitoringServiceTool.getInstance().getRegisteredUsers(deploymentUrl.toString(),adminAccountSid, adminAuthToken));
+
+        phone2.setAutoResponseOptionsRequests(true);
+    }
+    
+    @Test
+    public void registerUserAgentWithExtraParamsAnd408ToOptionsPing() throws ParseException, InterruptedException, InvalidArgumentException {
```
Comment (jaimecasero) [inline restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/ua/UserAgentManagerWithSBCTest.java]: "exception" category

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/ua/UserAgentManagerWithSBCTest.java [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+        logger.info("RequestEvent :"+requestEvent.getRequest().toString());
+        Response response = sipStack.getMessageFactory().createResponse(SIPResponse.OK, requestEvent.getRequest());
+        phone.sendReply(requestEvent, response);
+        Thread.sleep(1000);
+        // Clean up (Unregister).
+        assertTrue(phone.unregister(aliceContact, 0));
+    }
+
+    @Test
+    public void registerUserAgentWithExtraParamsAndOptionsPing() throws ParseException, InterruptedException {
```
Comment (jaimecasero) [inline restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/ua/UserAgentManagerWithSBCTest.java]: "alternate" category

File: restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/Call.java (hunk 1 above)
Comment (deruelle) [reply restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/Call.java]: it was present in the old code. didn't check that as it was out of scope

File: restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/ua/UserAgentManager.java (hunk 2 above)
Comment (deruelle) [reply restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/ua/UserAgentManager.java]: this is the default SIP ports for the given transports if not present in the SIP URI. We may want to have a util class for that as I think there is a couple of other places in RC or SIP Servlets that does this.

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/DialActionOrganizationSBCTest.java (hunk 3 above)
Comment (deruelle) [reply restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/DialActionOrganizationSBCTest.java]: I copied the DialActionOrganizationTest.java, not sure if it was present initially there. How would you convert that to parallel testing ?

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/DialActionOrganizationSBCTest.java (hunk 4 above)
Comment (deruelle) [reply restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/DialActionOrganizationSBCTest.java]: I'm not aware of what the exception category is. I copied the DialActionOrganizationTest.java, not sure if it was present initially there either.

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/DialActionOrganizationSBCTest.java (hunk 5 above)
Comment (deruelle) [reply restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/DialActionOrganizationSBCTest.java]: I copied the DialActionOrganizationTest.java, not sure if it was present initially there

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/ua/UserAgentManagerWithSBCTest.java (hunk 6 above)
Comment (deruelle) [reply restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/ua/UserAgentManagerWithSBCTest.java]: I copied the UserAgentManagerWithTest.java, not sure if it was present initially there

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/ua/UserAgentManagerWithSBCTest.java (hunk 7 above)
Comment (deruelle) [reply restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/ua/UserAgentManagerWithSBCTest.java]: I copied the UserAgentManagerWithTest.java, not sure if it was present initially there

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/ua/UserAgentManagerWithSBCTest.java (hunk 8 above)
Comment (deruelle) [reply restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/ua/UserAgentManagerWithSBCTest.java]: I copied the UserAgentManagerWithTest.java, not sure if it was present initially there

**Pull request merged**